File: matlab_code/test_csp.m
%%%%%%%%%% test_csp.m %%%%%%%%%%
function [ftr,fts,LABELS,ZTR,ZTS]=test_csp(TSDATA,TRDATA,TRLB,WCSP,params)
    Ktr=numel(TRDATA);
    for k=1:Ktr
        X=TRDATA{k};
        ZTR{k}=WCSP*X;
        ftr(k,:)=[log(var(ZTR{k},0,2)'/sum(var(ZTR{k},0,2)))];
    end
    Kts=numel(TSDATA);
    for k=1:Kts
        X=TSDATA{k};
        ZTS{k}=WCSP*X;
        fts(k,:)=[log(var(ZTS{k},0,2)'/sum(var(ZTS{k},0,2)))];
    end

    if(params.classifier=='LDA')
        MdlLinear = fitcdiscr(ftr, TRLB');
        LABELS = predict(MdlLinear, fts);
    elseif(params.classifier=='SVM')
        svmStruct = fitcsvm(ftr, TRLB);
        LABELS = ClassificationSVM(svmStruct, fts)';
end

File: python_code/test_csp.py
import numpy as np
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
from sklearn.svm import SVC

def test_csp(TSDATA, TRDATA, TRLB, WCSP, params):
    Ktr = len(TRDATA)
    ZTR = []
    ftr = []
    for k in range(Ktr):
        X = TRDATA[k]
        ZTR.append(WCSP @ X)
        ftr.append(np.log(np.var(ZTR[-1], axis=1) / np.sum(np.var(ZTR[-1], axis=1))))
    ftr = np.array(ftr)
    
    Kts = len(TSDATA)
    ZTS = []
    fts = []
    for k in range(Kts):
        X = TSDATA[k]
        ZTS.append(WCSP @ X)
        fts.append(np.log(np.var(ZTS[-1], axis=1) / np.sum(np.var(ZTS[-1], axis=1))))
    fts = np.array(fts)
    
    if params['classifier'] == 'LDA':
        clf = LDA()
        clf.fit(ftr, TRLB)
        LABELS = clf.predict(fts)
    elif params['classifier'] == 'SVM':
        clf = SVC()
        clf.fit(ftr, TRLB)
        LABELS = clf.predict(fts)
    
    return ftr, fts, LABELS, ZTR, ZTS
